### ISSUE TITLE
[Bug] bfloat8_b pack precision depends on which op produced it: typecast rounds once, matmul and the binary ops round twice

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#52155]
**Component / Area:** ttnn/eltwise, ttnn/matmul, tt_metal/jit_build

**Issue Type:** Bad Outputs

## Observed

The same `bfloat16 -> bfloat8_b` conversion gives two different answers depending on which op does it. `ttnn.typecast` rounds once; `ttnn.matmul`, `ttnn.add`, `ttnn.multiply` and `ttnn.clone` with a `bfloat8_b` output round twice — once into the packer's block-float intermediate, then again when mantissas shift to the block's shared exponent. Blackhole P300, 65536 random `bfloat16` values, each route in its own process:

| route | mean abs err | max abs err |
|---|---|---|
| `add` / `multiply` / `clone` -> `bfloat8_b` | **6.901e-3** | **2.344e-2** |
| `typecast` -> `bfloat8_b` | 5.840e-3 | 1.563e-2 |

15% lower mean error, exactly half the max error. `ttnn.matmul(a, b, dtype=ttnn.bfloat8_b)` also differs from `ttnn.typecast(ttnn.matmul(a, b), ttnn.bfloat8_b)` on 7716 / 65536 elements. Forcing `bfp8_pack_precise = true` for every kernel makes all four ops bit-identical to `typecast` and drops that 7716 to 0, so the flag is the whole of the effect.

## Expected

One rule, whichever one it is. The hardware documentation asks software to avoid exactly this double rounding — `tt-isa-documentation`, `WormholeB0/TensixTile/TensixCoprocessor/Packers/FormatConversion.md`:

> Two data format conversions happen during the pipeline of each packer: an early conversion just after reading datums from `Dst`, and a late conversion just before writing datums to `L1`. **To avoid double rounding, software should ensure that at least one conversion is an identity conversion, or ensure that both conversions are truncating.**

`bfp8_pack_precise = true` is exactly that: the early conversion becomes an identity (`Float16_b -> Float16_b`) and the single rounding happens in the late conversion. The default is the case the document warns against, so the two paths in the table above are not two tunings of one rule — one of them follows the documented guidance and the other does not. `typecast` (`typecast.cpp:42`) and a unary chain ending in `TYPECAST` (`unary.cpp:32`) follow the rule; binary and matmul never set the flag, and six experimental factories hardcode it to `false`. This is the unfinished half of PR #14822, whose review asked for binary ("can someone from MCW help to bring this to binary too?") and never got it; PR #46040 (2026-06) calls the flag "(which nobody uses)". A ttnn caller cannot ask for it either: no field on `DeviceComputeKernelConfig`, no environment variable, and the one binding that exists (`ttnn.ComputeConfigDescriptor`, for `ttnn.generic_op`) is unusable until #51204 lands.

## Steps to reproduce

```python
import torch, ttnn
d = ttnn.open_device(device_id=0)
torch.manual_seed(7)
a = torch.randn(1, 1, 256, 256, dtype=torch.bfloat16)
tt = lambda x: ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=d)

host    = ttnn.to_torch(ttnn.from_torch(a, dtype=ttnn.bfloat8_b, layout=ttnn.TILE_LAYOUT))
default = ttnn.to_torch(ttnn.add(tt(a), tt(torch.zeros_like(a)), dtype=ttnn.bfloat8_b))
precise = ttnn.to_torch(ttnn.typecast(tt(a), ttnn.bfloat8_b))

exact = a.to(torch.float32)
for name, t in (("default (add)", default), ("precise (typecast)", precise), ("host", host)):
    err = (t.to(torch.float32) - exact).abs()
    print(f"{name:20s} mean {err.mean():.6g}  max {err.max():.6g}  differs from host: {int((t != host).sum())}")
```

```
default (add)        0.00690097  max 0.0234375  differs from host: 14251
precise (typecast)   0.00583998  max 0.015625   differs from host: 6769
host                 0.00583998  max 0.015625   differs from host: 0
```

Compare anything that differs only in `bfp8_pack_precise` in separate processes: per #51204 two such kernels share one cache entry, and the second silently gets the first one's precision.

## What fixing it costs

Diffing the JIT artifacts for the same op built with the flag off and on, the only file that differs is `chlkc_pack_data_format.h`, and in it only `pack_src_format` (`6` = `Bfp8_b` becomes `5` = `Float16_b`). `pack_dst_format` is unchanged, so L1 still receives a `Bfp8_b` tile, and `chlkc_dst_accum_mode.h`, `chlkc_dst_sync_mode.h`, `chlkc_pack_tile_dims.h`, both `chlkc_unpack_*` and all three `chlkc_{unpack,math,pack}.cpp` are byte-identical.

- **Dst:** unchanged. `DST_ACCUM_MODE = false` and `DST_SYNC_MODE = SyncHalf` in both builds, so same element width and same tile capacity. The flag changes what the packer reads out of Dst, not Dst.
- **L1:** unchanged. Output tile stays 1088 B, no CB depends on the flag, and the pack kernel's `.text` is 1172 B with the flag off against **1168 B** with it on.
- **Throughput:** unchanged. In a loop where the packer is the critical path (64 cores, 512 `pack_tile` per core, 16 packs per `copy_tile`, no NOC traffic), cycles per packed tile at 1350 MHz are 61.3 with the flag off and **61.2** with it on, while a `float32`-output control in the same harness costs +23.4% — so the harness resolves a real pack-rate change and this is not one. At op level the pack thread does not move either (`add` 6557 -> 6552 cycles, `matmul` K=32 7760 -> 7764) and it is not the critical path; `BRISC` is. Whole-op deltas are within ±0.9% in both directions.

PR #14822 stated "half the packer bandwidth in the case of fp16" and #14032 added "shouldn't affect BH". Blackhole was not GA then, so this was measured rather than assumed.

Accuracy through a chain of 8 `ttnn.matmul` steps with `bfloat8_b` outputs, RMS-normalised between steps as a real model is, float32 reference on the same weights, seeds 0/1/2 agreeing:

| | flag off | flag on |
|---|---|---|
| rel L2 after 8 steps | 0.03047 | **0.02933** |
| PCC after 8 steps | 0.9995456 | **0.9995797** |

## What would settle it

1. **A decision on whether a `bfloat8_b` output should have the same precision whichever op produced it.** "No, the default stays" is a perfectly good answer — PR #14822 chose it for a reason, and Wormhole is where that reason was strongest, though the halved-bandwidth warning did not surface on WH at the shape I re-measured (see below). What is missing is that the answer is not written anywhere a caller can find, and that today the behaviour differs by op rather than by choice: `typecast` rounds once, `matmul` and the binary ops round twice, and nothing marks either as intended.

2. **If the answer is that it should:** the change is the same line `typecast.cpp:42` already has, applied in `binary_ng`, the matmul factories and `clone`. On Blackhole it costs nothing in throughput, Dst or L1 by the measurements above, and I have it with a consistency test if that is useful. It depends on #51204 — kernels differing only in this flag currently share a cache entry. Re-measured on Wormhole n300 today at `[1, 1, 256, 256]`: `add`/`mul`/`matmul` → `bfloat8_b` are all within ±1% of baseline in min-of-50-iterations wall-clock, so PR #14822's halved-bandwidth warning does not surface at that shape. Not swept; if a packer-bound configuration does regress on WH, an architecture gate is still the right response for that path.

3. **Either way, two documentation gaps.** The per-op rule is not stated anywhere: the only in-tree mention of the flag is `bfp_pack_precision_mode` in the Metal 2.0 header, and `DeviceComputeKernelConfig` does not carry it, so a ttnn caller can neither see nor choose. And `tech_reports/data_formats/data_formats.md` says block-float conversion breaks ties to even, which is true of the host packer and false of the hardware — the ISA document says BFP rounding is "deterministic round-to-nearest with ties away from zero", and on 1024 exact ties the device rounds 1008 away from zero while the host rounds 1008 to even. That divergence is deliberate (#15889 -> PR #16997 -> #17237, closed with the three host-vs-device typecast tests still skipped), but nothing says so where a caller would look.

## Environment

Blackhole P300, chip clock 1350 MHz. Code references verified at `b2e77906b9`. Measurements on a v0.67 release build whose `bfp8_pack_precise` branches in `data_format.cpp` are identical to it; the accuracy table also reproduces on `25891d3` and `66d22c1f3a9`. ISA quotes from `tt-isa-documentation@main`, which has no `Packers/` page under `BlackholeA0/`.

**Not established:** WH re-measurement (2026-08-14, tyo-qb01, n300) confirms the bug reproduces at the same counts as BH (`add` 7553/65536, `matmul` 7716/65536, mean 0.0069 vs typecast's 0.0058), the fix works, and PR #14822's halved-bandwidth warning does not surface in wall-clock for add/mul/matmul → `bfloat8_b` at `[1, 1, 256, 256]`. What still is not established: end-to-end model trace, WH packer-bound shape sweep, cycle-level `PACK_ISOLATE` timing, and whether the omission for binary and matmul was a decision or an oversight. In `models/`, 52 single-line call sites ask a device op for a `bfloat8_b` output (15 of them `ttnn.typecast`, so already precise); the ~2000 other `bfloat8_b` mentions are weights and KV-cache dtypes converted on the host, where this flag is irrelevant.

## Related

#14032 / PR #14822 (the original report and the flag) · #46040 ("which nobody uses") · #51204 (the hash omission, a prerequisite) · #15889 / PR #16997 / #17237 (the host tie rule and the skipped tests) · #51557 (NaN packed as Inf, a different mechanism) · #52141 (a compile-time decision the caller cannot see changing numerics, one layer up)
